File: web_app/ofdmEngine.js
/**
 * OFDM Transceiver DSP Engine (Client-Side JS)
 * Project: OFDM-Based Wireless Communication System with Channel Estimation and Equalization
 */

class OFDMTransceiverJS {
  constructor(options = {}) {
    this.nFFT = options.nFFT || 64;
    this.cpLength = options.cpLength || 16;
    this.modulation = (options.modulation || 'QPSK').toUpperCase();
    this.pilotSpacing = options.pilotSpacing || 4;
    this.bitsPerSymbol = this.modulation === 'BPSK' ? 1 : 2;

    const all = Array.from({ length: this.nFFT }, (_, i) => i);
    this.pilotIndices = all.filter(i => i > 0 && i % this.pilotSpacing === 0);
    this.dataIndices = all.filter(i => i > 0 && !this.pilotIndices.includes(i));
    
    this.nPilots = this.pilotIndices.length;
    this.nData = this.dataIndices.length;
    
    this.pilotValue = this.modulation === 'BPSK' 
      ? { re: 1.0, im: 0.0 } 
      : { re: 1.0 / Math.SQRT2, im: 1.0 / Math.SQRT2 };

    const rawPdp = [1.0, Math.exp(-1.0 / 1.5), Math.exp(-2.0 / 1.5), Math.exp(-3.0 / 1.5)];
    const sumPdp = rawPdp.reduce((a, b) => a + b, 0);
    this.pdp = rawPdp.map(v => v / sumPdp);
  }

  generateBits(nSymbols) {
    const totalBits = nSymbols * this.nData * this.bitsPerSymbol;
    const bits = new Uint8Array(totalBits);
    for (let i = 0; i < totalBits; i++) {
      bits[i] = Math.random() < 0.5 ? 0 : 1;
    }
    return bits;
  }

  mapBitsToSymbols(bits) {
    const symbols = [];
    if (this.modulation === 'BPSK') {
      for (let i = 0; i < bits.length; i++) {
        symbols.push({ re: bits[i] === 0 ? 1.0 : -1.0, im: 0.0 });
      }
    } else {
      for (let i = 0; i < bits.length; i += 2) {
        const b0 = bits[i];
        const b1 = bits[i + 1];
        const re = (b0 === 0 ? 1.0 : -1.0) / Math.SQRT2;
        const im = (b1 === 0 ? 1.0 : -1.0) / Math.SQRT2;
        symbols.push({ re, im });
      }
    }
    return symbols;
  }

  demodulateSymbolsToBits(symbols) {
    const bits = [];
    if (this.modulation === 'BPSK') {
      for (let i = 0; i < symbols.length; i++) {
        bits.push(symbols[i].re < 0 ? 1 : 0);
      }
    } else {
      for (let i = 0; i < symbols.length; i++) {
        bits.push(symbols[i].re < 0 ? 1 : 0);
        bits.push(symbols[i].im < 0 ? 1 : 0);
      }
    }
    return new Uint8Array(bits);
  }

  ifft(freqFrame) {
    const N = freqFrame.length;
    const conjInput = freqFrame.map(c => ({ re: c.re, im: -c.im }));
    const fftOutput = this.fft(conjInput);
    return fftOutput.map(c => ({
      re: (c.re / N) * Math.sqrt(N),
      im: (-c.im / N) * Math.sqrt(N)
    }));
  }

  fft(x) {
    const N = x.length;
    if (N <= 1) return x;
    if ((N & (N - 1)) !== 0) {
      return this.dft(x);
    }
    const even = this.fft(x.filter((_, i) => i % 2 === 0));
    const odd = this.fft(x.filter((_, i) => i % 2 === 1));

    const X = new Array(N);
    for (let k = 0; k < N / 2; k++) {
      const angle = (-2 * Math.PI * k) / N;
      const twiddle = { re: Math.cos(angle), im: Math.sin(angle) };
      const ok = {
        re: odd[k].re * twiddle.re - odd[k].im * twiddle.im,
        im: odd[k].re * twiddle.im + odd[k].im * twiddle.re
      };
      X[k] = { re: even[k].re + ok.re, im: even[k].im + ok.im };
      X[k + N / 2] = { re: even[k].re - ok.re, im: even[k].im - ok.im };
    }
    return X;
  }

  dft(x) {
    const N = x.length;
    const X = new Array(N);
    for (let k = 0; k < N; k++) {
      let sumRe = 0, sumIm = 0;
      for (let n = 0; n < N; n++) {
        const angle = (-2 * Math.PI * k * n) / N;
        sumRe += x[n].re * Math.cos(angle) - x[n].im * Math.sin(angle);
        sumIm += x[n].re * Math.sin(angle) + x[n].im * Math.cos(angle);
      }
      X[k] = { re: sumRe, im: sumIm };
    }
    return X;
  }

  transmit(dataSymbols) {
    const nSymbols = Math.floor(dataSymbols.length / this.nData);
    const txTimeFrames = [];
    const txFreqFrames = [];

    for (let s = 0; s < nSymbols; s++) {
      const freqFrame = new Array(this.nFFT).fill(null).map(() => ({ re: 0, im: 0 }));
      for (let d = 0; d < this.nData; d++) {
        freqFrame[this.dataIndices[d]] = dataSymbols[s * this.nData + d];
      }
      for (let p = 0; p < this.nPilots; p++) {
        freqFrame[this.pilotIndices[p]] = { ...this.pilotValue };
      }
      freqFrame[0] = { re: 0, im: 0 };
      txFreqFrames.push(freqFrame);

      const timeFrame = this.ifft(freqFrame);
      const cp = timeFrame.slice(this.nFFT - this.cpLength);
      const timeWithCp = cp.concat(timeFrame);
      txTimeFrames.push(timeWithCp);
    }

    const txSignal = [];
    txTimeFrames.forEach(f => f.forEach(sample => txSignal.push(sample)));
    return { txSignal, txFreqFrames, txTimeFrames };
  }

  generateRayleighChannel() {
    const taps = [];
    for (let i = 0; i < this.pdp.length; i++) {
      const sigma = Math.sqrt(this.pdp[i] / 2.0);
      const u1 = Math.max(1e-10, Math.random());
      const u2 = Math.random();
      const z0 = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
      const z1 = Math.sqrt(-2.0 * Math.log(u1)) * Math.sin(2.0 * Math.PI * u2);
      taps.push({ re: z0 * sigma, im: z1 * sigma });
    }
    return taps;
  }

  passChannel(txSignal, channelTaps, snrDb, channelType = 'RAYLEIGH') {
    let rxConv = [];
    let actualTaps = channelTaps;

    if (channelType.toUpperCase() === 'AWGN') {
      rxConv = txSignal.map(s => ({ ...s }));
      actualTaps = [{ re: 1.0, im: 0.0 }];
    } else {
      const N = txSignal.length;
      const L = channelTaps.length;
      rxConv = new Array(N).fill(null).map(() => ({ re: 0, im: 0 }));

      for (let n = 0; n < N; n++) {
        let reSum = 0, imSum = 0;
        for (let l = 0; l < L; l++) {
          if (n - l >= 0) {
            const x = txSignal[n - l];
            const h = channelTaps[l];
            reSum += x.re * h.re - x.im * h.im;
            imSum += x.re * h.im + x.im * h.re;
          }
        }
        rxConv[n] = { re: reSum, im: imSum };
      }
    }

    let pSignal = 0;
    rxConv.forEach(s => pSignal += s.re * s.re + s.im * s.im);
    pSignal /= rxConv.length;

    const snrLinear = Math.pow(10, snrDb / 10.0);
    const noisePower = pSignal / Math.max(1e-12, snrLinear);
    const sigmaNoise = Math.sqrt(noisePower / 2.0);

    const rxNoisy = rxConv.map(s => {
      const u1 = Math.max(1e-10, Math.random());
      const u2 = Math.random();
      const n0 = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2) * sigmaNoise;
      const n1 = Math.sqrt(-2.0 * Math.log(u1)) * Math.sin(2.0 * Math.PI * u2) * sigmaNoise;
      return { re: s.re + n0, im: s.im + n1 };
    });

    return { rxNoisy, actualTaps, noisePower };
  }

  receive(rxSignal, nSymbols) {
    const frameLen = this.nFFT + this.cpLength;
    const rxFreqFrames = [];

    for (let s = 0; s < nSymbols; s++) {
      const frameWithCp = rxSignal.slice(s * frameLen, (s + 1) * frameLen);
      if (frameWithCp.length < frameLen) break;
      const timeFrame = frameWithCp.slice(this.cpLength);
      
      const freqRaw = this.fft(timeFrame);
      const freqFrame = freqRaw.map(c => ({
        re: c.re / Math.sqrt(this.nFFT),
        im: c.im / Math.sqrt(this.nFFT)
      }));
      rxFreqFrames.push(freqFrame);
    }
    return rxFreqFrames;
  }

  estimateChannel(rxFreqFrames, method = 'LS', noisePower = 1e-3) {
    const HEst = [];
    const nSymbols = rxFreqFrames.length;

    for (let s = 0; s < nSymbols; s++) {
      const rxFreq = rxFreqFrames[s];
      const hPilots = [];

      for (let p = 0; p < this.nPilots; p++) {
        const pIdx = this.pilotIndices[p];
        const Y = rxFreq[pIdx];
        const X = this.pilotValue;
        const xMagSq = X.re * X.re + X.im * X.im;
        const hLS = {
          re: (Y.re * X.re + Y.im * X.im) / xMagSq,
          im: (Y.im * X.re - Y.re * X.im) / xMagSq
        };
        hPilots.push({ idx: pIdx, ...hLS });
      }

      const hFrame = new Array(this.nFFT);
      for (let k = 0; k < this.nFFT; k++) {
        let left = hPilots[0];
        let right = hPilots[hPilots.length - 1];

        for (let p = 0; p < hPilots.length - 1; p++) {
          if (k >= hPilots[p].idx && k <= hPilots[p + 1].idx) {
            left = hPilots[p];
            right = hPilots[p + 1];
            break;
          }
        }

        let interpRe, interpIm;
        if (left.idx === right.idx) {
          interpRe = left.re;
          interpIm = left.im;
        } else {
          const alpha = (k - left.idx) / (right.idx - left.idx);
          interpRe = left.re + alpha * (right.re - left.re);
          interpIm = left.im + alpha * (right.im - left.im);
        }

        if (method.toUpperCase() === 'MMSE') {
          const magSq = interpRe * interpRe + interpIm * interpIm;
          const factor = magSq / (magSq + noisePower);
          interpRe *= factor;
          interpIm *= factor;
        }

        hFrame[k] = { re: interpRe, im: interpIm };
      }
      HEst.push(hFrame);
    }
    return HEst;
  }

  equalize(rxFreqFrames, HEst, method = 'ZF', snrDb = 20) {
    const snrLinear = Math.pow(10, snrDb / 10.0);
    const equalizedDataSymbols = [];
    const equalizedFreqFrames = [];

    for (let s = 0; s < rxFreqFrames.length; s++) {
      const Y = rxFreqFrames[s];
      const H = HEst[s];
      const eqFrame = new Array(this.nFFT);

      for (let k = 0; k < this.nFFT; k++) {
        const yk = Y[k];
        const hk = H[k];
        const hMagSq = hk.re * hk.re + hk.im * hk.im;

        let xHat = { re: 0, im: 0 };
        if (method.toUpperCase() === 'ZF') {
          const denom = Math.max(1e-10, hMagSq);
          xHat = {
            re: (yk.re * hk.re + yk.im * hk.im) / denom,
            im: (yk.im * hk.re - yk.re * hk.im) / denom
          };
        } else {
          const denom = hMagSq + (1.0 / snrLinear);
          xHat = {
            re: (yk.re * hk.re + yk.im * hk.im) / denom,
            im: (yk.im * hk.re - yk.re * hk.im) / denom
          };
        }
        eqFrame[k] = xHat;
      }

      equalizedFreqFrames.push(eqFrame);
      for (let d = 0; d < this.nData; d++) {
        equalizedDataSymbols.push(eqFrame[this.dataIndices[d]]);
      }
    }
    return { equalizedDataSymbols, equalizedFreqFrames };
  }

  computeBER(txBits, rxBits) {
    let bitErrors = 0;
    const len = Math.min(txBits.length, rxBits.length);
    for (let i = 0; i < len; i++) {
      if (txBits[i] !== rxBits[i]) bitErrors++;
    }
    return { ber: bitErrors / len, bitErrors, totalBits: len };
  }
}

if (typeof window !== 'undefined') {
  window.OFDMTransceiverJS = OFDMTransceiverJS;
}

File: ofdm_engine.py
"""
OFDM Transceiver DSP Engine - Sig2Sig Implementation
Provides complete end-to-end signal processing for OFDM communications:
- Bit Generation & Modulation (BPSK, QPSK with Gray coding)
- Comb-type Pilot Insertion & Subcarrier Allocation
- IFFT & Cyclic Prefix (CP) addition/removal
- Multipath Rayleigh Fading Channel & AWGN Noise Injection
- Least-Squares (LS) & MMSE Channel Estimation
- Zero-Forcing (ZF) & MMSE Equalization
- Hard Decision Demodulation & BER Computation
"""

import numpy as np
from scipy import interpolate

class OFDMTransceiver:
    def __init__(self, n_fft=64, cp_length=16, modulation='QPSK', pilot_spacing=4, channel_taps=None):
        """
        Parameters:
        -----------
        n_fft : int
            Number of FFT subcarriers (default 64)
        cp_length : int
            Cyclic prefix length in samples (default 16)
        modulation : str
            'BPSK' or 'QPSK'
        pilot_spacing : int
            Spacing between pilot subcarriers (comb-type pilots)
        channel_taps : array-like or None
            Channel impulse response taps. If None, default 4-tap Rayleigh profile used.
        """
        self.n_fft = n_fft
        self.cp_length = cp_length
        self.modulation = modulation.upper()
        self.pilot_spacing = pilot_spacing
        
        # Subcarrier allocation (DC at 0, Guard bands at edges, Pilots every pilot_spacing)
        # We reserve subcarrier 0 as DC null. Edge subcarriers can be active.
        all_indices = np.arange(self.n_fft)
        
        # Pilot indices: 0 is DC, so pilots at 1, 1+spacing, ...
        self.pilot_indices = all_indices[1::self.pilot_spacing]
        self.data_indices = np.array([i for i in all_indices if i not in self.pilot_indices and i != 0])
        
        self.n_pilots = len(self.pilot_indices)
        self.n_data = len(self.data_indices)
        
        # Fixed pilot values (known to Tx and Rx)
        self.pilot_value = 1.0 + 0.0j if self.modulation == 'BPSK' else (1.0 + 1.0j) / np.sqrt(2)
        
        # Bits per symbol
        self.bits_per_symbol = 1 if self.modulation == 'BPSK' else 2
        
        # Default channel impulse response if not provided
        if channel_taps is None:
            # 4-tap Rayleigh fading channel with exponential power delay profile
            pdp = np.exp(-np.arange(4) / 1.5)
            pdp /= np.sum(pdp)  # Normalize total channel power to 1
            self.pdp = pdp
        else:
            self.pdp = np.array(channel_taps) / np.sqrt(np.sum(np.abs(channel_taps)**2))
            
    def generate_bits(self, n_ofdm_symbols):
        """Generate random binary sequence for given number of OFDM symbols."""
        total_bits = n_ofdm_symbols * self.n_data * self.bits_per_symbol
        return np.random.randint(0, 2, total_bits)
        
    def map_bits_to_symbols(self, bits):
        """Map bits to BPSK or QPSK complex constellation symbols with Gray coding."""
        if self.modulation == 'BPSK':
            # 0 -> +1, 1 -> -1
            symbols = 1.0 - 2.0 * bits.astype(float)
            return symbols.astype(complex)
        elif self.modulation == 'QPSK':
            # Group into 2-bit pairs (Gray mapping)
            bits_reshaped = bits.reshape(-1, 2)
            # 00 -> (1+j)/sqrt(2), 01 -> (-1+j)/sqrt(2), 11 -> (-1-j)/sqrt(2), 10 -> (1-j)/sqrt(2)
            i_val = 1.0 - 2.0 * bits_reshaped[:, 0]
            q_val = 1.0 - 2.0 * bits_reshaped[:, 1]
            symbols = (i_val + 1j * q_val) / np.sqrt(2.0)
            return symbols
        else:
            raise ValueError(f"Unsupported modulation: {self.modulation}")
            
    def demodulate_symbols_to_bits(self, symbols):
        """Hard decision demodulation from complex symbols back to bits."""
        if self.modulation == 'BPSK':
            # Re(symbol) > 0 -> bit 0, else bit 1
            return (np.real(symbols) < 0.0).astype(int)
        elif self.modulation == 'QPSK':
            bits_i = (np.real(symbols) < 0.0).astype(int)
            bits_q = (np.imag(symbols) < 0.0).astype(int)
            bits = np.column_stack((bits_i, bits_q)).ravel()
            return bits

    def transmit(self, data_symbols):
        """
        Build OFDM frame in frequency domain, apply IFFT, add CP.
        data_symbols shape: (n_ofdm_symbols, n_data)
        Returns:
        --------
        tx_signal : 1D complex array of time domain OFDM signal
        tx_freq_frames : 2D complex array of frequency domain subcarriers (n_symbols, n_fft)
        """
        n_symbols = len(data_symbols)
        tx_freq_frames = np.zeros((n_symbols, self.n_fft), dtype=complex)
        
        # Populate subcarriers
        tx_freq_frames[:, self.data_indices] = data_symbols
        tx_freq_frames[:, self.pilot_indices] = self.pilot_value
        tx_freq_frames[:, 0] = 0.0  # DC null
        
        # IFFT per symbol (scaling by sqrt(n_fft) for energy preservation)
        tx_time_frames = np.fft.ifft(tx_freq_frames, axis=1) * np.sqrt(self.n_fft)
        
        # Cyclic Prefix addition
        cp = tx_time_frames[:, -self.cp_length:]
        tx_time_with_cp = np.hstack((cp, tx_time_frames))
        
        # Serialize for transmission
        tx_signal = tx_time_with_cp.ravel()
        return tx_signal, tx_freq_frames

    def generate_rayleigh_channel(self, num_taps=None):
        """Generate random Rayleigh fading taps sampled from complex Gaussian distribution."""
        if num_taps is None:
            num_taps = len(self.pdp)
        sigma = np.sqrt(self.pdp / 2.0)
        taps = (np.random.normal(0, sigma) + 1j * np.random.normal(0, sigma))
        return taps

    def pass_channel(self, tx_signal, channel_taps, snr_db, channel_type='RAYLEIGH'):
        """
        Pass signal through wireless channel (Multipath Rayleigh Fading or AWGN) and add noise.
        """
        if channel_type.upper() == 'AWGN':
            rx_signal = tx_signal.copy()
            actual_channel = np.array([1.0])
        else:  # RAYLEIGH
            rx_signal = np.convolve(tx_signal, channel_taps, mode='full')[:len(tx_signal)]
            actual_channel = channel_taps

        # Calculate noise variance for given SNR dB
        # SNR_linear = P_signal / P_noise
        signal_power = np.mean(np.abs(rx_signal)**2)
        snr_linear = 10.0 ** (snr_db / 10.0)
        noise_power = signal_power / snr_linear
        
        # Complex AWGN noise generator
        noise = np.sqrt(noise_power / 2.0) * (np.random.randn(*rx_signal.shape) + 1j * np.random.randn(*rx_signal.shape))
        rx_signal_noisy = rx_signal + noise
        
        return rx_signal_noisy, actual_channel, noise_power

    def receive(self, rx_signal, n_symbols):
        """
        De-serialize, remove CP, and perform FFT.
        Returns:
        --------
        rx_freq_frames : 2D complex array (n_symbols, n_fft)
        """
        frame_len = self.n_fft + self.cp_length
        rx_time_with_cp = rx_signal[:n_symbols * frame_len].reshape(n_symbols, frame_len)
        
        # CP removal
        rx_time_frames = rx_time_with_cp[:, self.cp_length:]
        
        # FFT per symbol
        rx_freq_frames = np.fft.fft(rx_time_frames, axis=1) / np.sqrt(self.n_fft)
        return rx_freq_frames

    def estimate_channel(self, rx_freq_frames, method='LS', noise_power=1e-3):
        """
        Perform Channel Estimation using pilots.
        Methods:
        --------
        - 'LS': Least Squares estimation at pilot locations + Linear interpolation.
        - 'MMSE': Minimum Mean Square Error estimation.
        """
        n_symbols = len(rx_freq_frames)
        H_est = np.zeros((n_symbols, self.n_fft), dtype=complex)
        
        for i in range(n_symbols):
            rx_pilots = rx_freq_frames[i, self.pilot_indices]
            
            # LS estimation at pilot subcarriers
            H_pilots_ls = rx_pilots / self.pilot_value
            
            if method.upper() == 'LS':
                # Interpolate over all subcarriers
                # Handle boundaries by duplicating edge pilots
                pilots_idx = self.pilot_indices
                H_p_real = np.real(H_pilots_ls)
                H_p_imag = np.imag(H_pilots_ls)
                
                f_real = interpolate.interp1d(pilots_idx, H_p_real, kind='linear', fill_value='extrapolate')
                f_imag = interpolate.interp1d(pilots_idx, H_p_imag, kind='linear', fill_value='extrapolate')
                
                all_idx = np.arange(self.n_fft)
                H_est[i, :] = f_real(all_idx) + 1j * f_imag(all_idx)

            elif method.upper() == 'MMSE':
                # MMSE estimate derived from LS pilots with noise variance smoothing
                pilots_idx = self.pilot_indices
                H_p_real = np.real(H_pilots_ls)
                H_p_imag = np.imag(H_pilots_ls)
                
                f_real = interpolate.interp1d(pilots_idx, H_p_real, kind='linear', fill_value='extrapolate')
                f_imag = interpolate.interp1d(pilots_idx, H_p_imag, kind='linear', fill_value='extrapolate')
                
                all_idx = np.arange(self.n_fft)
                H_ls_interp = f_real(all_idx) + 1j * f_imag(all_idx)
                
                # Frequency-domain MMSE smoothing filter
                snr_inv = noise_power
                H_est[i, :] = H_ls_interp * (np.abs(H_ls_interp)**2 / (np.abs(H_ls_interp)**2 + snr_inv))
            else:
                raise ValueError(f"Unknown estimation method: {method}")
                
        return H_est

    def equalize(self, rx_freq_frames, H_est, method='ZF', snr_db=20):
        """
        Channel Equalization.
        Methods:
        --------
        - 'ZF': Zero Forcing (X_hat = Y / H_est)
        - 'MMSE': Minimum Mean Square Error (X_hat = Y * H_est* / (|H_est|^2 + 1/SNR))
        """
        snr_linear = 10.0 ** (snr_db / 10.0)
        
        if method.upper() == 'ZF':
            # Avoid division by zero with small epsilon
            H_safe = np.where(np.abs(H_est) < 1e-12, 1e-12, H_est)
            equalized_freq = rx_freq_frames / H_safe
        elif method.upper() == 'MMSE':
            # MMSE equalizer weight: W = H* / (|H|^2 + 1/SNR)
            denom = np.abs(H_est)**2 + (1.0 / snr_linear)
            equalized_freq = rx_freq_frames * np.conj(H_est) / denom
        else:
            raise ValueError(f"Unknown equalization method: {method}")
            
        # Extract data subcarriers
        equalized_data_symbols = equalized_freq[:, self.data_indices]
        return equalized_data_symbols, equalized_freq

    def compute_ber(self, tx_bits, rx_bits):
        """Compute Bit Error Rate (BER)."""
        bit_errors = np.sum(tx_bits != rx_bits)
        ber = bit_errors / len(tx_bits)
        return ber, bit_errors
